### PR TITLE
feat: adds dynamic routing.

### PR DESCRIPTION
**Pull request by atulep**
Adds dynamic routing feature.

### CODE REVIEW

**Comment by atulep** [comment]
This PR is almost ready, but misses the following:

1. Associated changes to `ads-templates`.
2. Updated integration golden files.

I will make changes for 1) after I get an approval of the method used in regular templates. As far as 2), I'm currently running into `bazel` build issues that I need to debug.

File: tests/unit/schema/wrappers/test_routing.py
Comment (software-dov) [inline tests/unit/schema/wrappers/test_routing.py:105]: This is a little hard to eyeball parse. Can you reformat it using black?

File: gapic/schema/wrappers.py
Comment (software-dov) [inline gapic/schema/wrappers.py:851]: Is this code path ever tested?

**Review by software-dov — COMMENTED** [review]
One worry: there _are_ APIs that use separators other than `/` in their path templates. I think that the logic here may need to support that.

File: gapic/schema/wrappers.py
Comment (atulep) [reply gapic/schema/wrappers.py:854]: I don't know a good variable name.

File: gapic/templates/tests/unit/gapic/%name_%version/%sub/test_%service.py.j2
Comment (atulep) [reply gapic/templates/tests/unit/gapic/%name_%version/%sub/test_%service.py.j2:730]: No, ideally, we want to assert correct x-goog-parameters are set. However, it's impossible to do it since the headers are set using complicated process difficult to mock.

File: gapic/schema/wrappers.py
Comment (atulep) [reply gapic/schema/wrappers.py:1056]: It's already there - tests/fragments/test_dynamic_routing.proto.

File: gapic/schema/wrappers.py
Comment (atulep) [reply gapic/schema/wrappers.py:851]: Yes, the coverage is 100%

File: gapic/schema/wrappers.py
Comment (atulep) [reply gapic/schema/wrappers.py]: Done

File: gapic/schema/wrappers.py
Comment (atulep) [reply gapic/schema/wrappers.py]: Done

File: tests/unit/schema/wrappers/test_routing.py
Comment (atulep) [reply tests/unit/schema/wrappers/test_routing.py]: I ran `black`, but it looks like style_check CI test fails. So, looks like the file needs to be formatted using `find gapic tests -name "*.py" -not -path 'tests/**/goldens/*' | xargs autopep8 --in-place`

**Comment by atulep** [comment]
> there _are_ APIs that use separators other than `/` in their path templates. I think that the logic here may need to support that.

Can you elaborate? I think path template in routing parameter always uses '/'.

File: tests/unit/schema/wrappers/test_method.py
Comment (software-dov) [inline tests/unit/schema/wrappers/test_method.py:303]: I don't think this check is necessary; the check above tests equality, which should check the type since you haven't overridden `__equal__`.

**Review by software-dov — APPROVED** [review]

File: gapic/schema/wrappers.py
Comment (atulep) [reply gapic/schema/wrappers.py:778]: There must be 0, 1 or 2 segment IDs. This corresponds to one named segment (i.e. 0 - no segments, 1 or 2 - one named segment). There can only be one named segment based on the requirements of the project. I raise a `ValueError` if that's not the case at the very beginning of execution.

File: gapic/schema/wrappers.py
Comment (atulep) [reply gapic/schema/wrappers.py:828]: Done.

File: tests/unit/schema/wrappers/test_method.py
Comment (atulep) [reply tests/unit/schema/wrappers/test_method.py:303]: Done

**Pull request merged**
